### PR TITLE
kriticism/server forking and training loop

### PR DESCRIPTION
**Pull request by kriticism**
# Description 

This PR adds an endpoint `/board/create`. Hitting theendpoint creates a fork and runs the model training on it. Openrec is needed to use this.

# Test/Benchmark Outputs
Tested the try/except block manually. Old endpoints are working correctly too.

# Checklist:
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [ ] Related Benchmarks have been performed, if applicable.
- [x] I have a plan to revert my changes.
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation, if applicable.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -102,3 +103,10 @@ venv.bak/
 
 # mypy
 .mypy_cache/
+
+# index checkpoints | to be removed after POC is complete
```
Comment (mohit-chawla) [inline .gitignore]: Why are they to be removed after PoC is complete? 
We should never push model-data to repo?

Comment (kriticism) [reply .gitignore]: Sounds good, will change the comment

File: board/model_manager.py [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+        self.test_sampler = test_sampler
+        self.evaluation_metric = evaluation_metric
+        self.path_to_dataset = path_to_dataset
+
+        self.model_id = '100xxks'
+        print("ModelManager init generated model id=", self.model_id)
+
+        
+    def sample_data_and_train(self):
+        print("sample_data_and_train called, pid =", os.getpid())
```
Comment (mohit-chawla) [inline board/model_manager.py]: nit: Can you use https://docs.python.org/2/library/logging.html#logging.Logger.debug here?

File: board/model_manager.py [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+        print("-------- sample_data_and_train starts --------")
```
Comment (mohit-chawla) [inline board/model_manager.py]: same

File: board/model_manager.py [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+        total_users = 0
+        interactions_count = 0
+        with open(os.path.dirname(os.path.abspath(__file__))+self.path_to_dataset, 'r') as fin:
+            for line in fin:
+                interactions_count += int(line.split()[0])
+                total_users += 1
+        print("############ collecting data.. ############")
```
Comment (mohit-chawla) [inline board/model_manager.py]: same

File: board/model_manager.py [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+class ModelManager:
+    def __init__(self, recommender_name, train_sampler, val_sampler, test_sampler, evaluation_metric, path_to_dataset):
+        self.recommender_name = recommender_name
+        self.train_sampler = train_sampler
+        self.val_sampler = val_sampler
+        self.test_sampler = test_sampler
+        self.evaluation_metric = evaluation_metric
+        self.path_to_dataset = path_to_dataset
+
+        self.model_id = '100xxks'
```
Comment (mohit-chawla) [inline board/model_manager.py]: Lets add an issue for this, so that we know what is left to do.

File: board/model_manager.py [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+                        test_structured_arr[next_user_id] = (next_user_id, 
+                                                            map_to_item_id[item])
+                    else:
+                        train_structured_arr[interaction_ind] = (next_user_id, 
+                                                                map_to_item_id[item])
+                        interaction_ind += 1
+                next_user_id += 1
+
+
+        print("############ instantiating dataset.. ############")
```
Comment (mohit-chawla) [inline board/model_manager.py]: same

File: board/model_manager.py [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+                            total_items=len(map_to_item_id),
+                            num_negatives=500,
+                            name='Val')
+        test_dataset = Dataset(raw_data=test_structured_arr,
+                            total_users=total_users,
+                            total_items=len(map_to_item_id),
+                            num_negatives=500,
+                            name='Test')
+
+        print("############ instantiating Samplers.. ############")
```
Comment (mohit-chawla) [inline board/model_manager.py]: same

File: board/model_manager.py (hunk 5 above)
Comment (mohit-chawla) [reply board/model_manager.py]: I added an issue

File: board/model_manager.py [hunk 8]
```diff
@@ -10,21 +13,25 @@ def __init__(self, recommender_name, train_sampler, val_sampler, test_sampler, e
         self.test_sampler = test_sampler
         self.evaluation_metric = evaluation_metric
         self.path_to_dataset = path_to_dataset
+        
+        logging.basicConfig(format=FORMAT)
+        self.logger = logging.getLogger('modelmanager')
```
Comment (mohit-chawla) [inline board/model_manager.py]: We should have memory efficient logging. Need to understand- 
1. Internals of python logger
2. Decide how many unique instances of logger to use for the project (if we need more than one).
Approving for now. Pl create an enhancement issue `Mem efficient logging`

Comment (mohit-chawla) [reply board/model_manager.py]: Lets use `logging.getLogger(__name__)` for now

**Pull request merged**